src/main/java/com/amplitude/Retry.java

```diff
@@ -0,0 +1,268 @@
+                }
+              }
+              eventsInRetry.addAndGet(-eventCount);
+            });
+    retryThread.start();
+  }
+
+  // Call this function if event not in current Retry list.
+  private static void onEventsError(List<Event> events, Response response, String apiKey) {
+    List<Event> eventsToRetry = events;
+    // Filter invalid event out based on the response code.
+    if (response.status == Status.RATELIMIT && response.RateLimitBody != null) {
+      // JSONObject deviceId as key, number as value
+      JSONObject exceededDailyQuotaUsers =
+          response.RateLimitBody.getJSONObject("exceededDailyQuotaUsers");
+      JSONObject exceededDailyQuotaDevices =
+          response.RateLimitBody.getJSONObject("exceededDailyQuotaDevices");
+      eventsToRetry =
+          events.stream()
+              .filter(
+                  (event ->
+                      !(event.userId != null && exceededDailyQuotaUsers.has(event.userId))
+                          && !(event.deviceId != null
+                              && exceededDailyQuotaDevices.has(event.deviceId))))
+              .collect(Collectors.toList());
+    } else if (response.status == Status.INVALID) {
+      if ((response.InvalidRequestBody != null
+              && response.InvalidRequestBody.has("missingField")
+              && response.InvalidRequestBody.getString("missingField").length() > 0)
+          || events.size() == 1) {
+        // Return early if there's an issue with the entire payload
+        // or if there's only one event and its invalid
+        return;
+      } else if (response.InvalidRequestBody != null) {
+        // Filter out invalid events id
+        int[] invalidEventIndices = collectInvalidEventIndices(response);
+        eventsToRetry =
+            IntStream.range(0, events.size())
+                .filter(i -> Arrays.binarySearch(invalidEventIndices, i) < 0)
+                .mapToObj(events::get)
+                .collect(Collectors.toList());
+      }
+    } else if (response.status == Status.SUCCESS) {
+      return;
+    }
+
+    for (Event event : eventsToRetry) {
+      String userId = (event.userId != null) ? event.userId : "";
+      String deviceId = (event.deviceId != null) ? event.deviceId : "";
+      if (userId.length() > 0 || deviceId.length() > 0) {
+        Map<String, List<Event>> deviceToBufferMap = idToBuffer.get(userId);
+        if (deviceToBufferMap == null) {
+          deviceToBufferMap = new ConcurrentHashMap<String, List<Event>>();
+          idToBuffer.put(userId, deviceToBufferMap);
+        }
+        List<Event> retryBuffer = deviceToBufferMap.get(deviceId);
+        if (retryBuffer == null) {
+          retryBuffer = new ArrayList<Event>();
+          deviceToBufferMap.put(deviceId, retryBuffer);
+          retryEventsOnLoop(userId, deviceId, apiKey);
+        }
+        eventsInRetry.incrementAndGet();
+        retryBuffer.add(event);
+      }
+    }
+    idToBuffer.forEach(
```

Comment:
Curious why this is preferred comparing to store the userId and deviceId information in a list then looping it? And I guess ConcurrentHashMap helps to avoid ConcurrentModificationException here? If we have to have two thread in onEventsError, would that kick off two threads for same user and deviceId or run before one loop in one thread ends? I guess if the events happen to share the user and device id we may hit this case as well, however if they don't would this just happens to create that case more?